Feature/export

AdityaKhatri opened the pull request:
- Depends on [135](https://github.com/collective-service/ifrc-gates-backend/pull/135)

## Fix
- https://github.com/collective-service/ifrc-gates-dashboard/issues/259
- https://github.com/collective-service/ifrc-gates-dashboard/issues/334
- https://github.com/collective-service/ifrc-gates-dashboard/issues/335

## Changes

* Add export functionality

## This PR doesn't introduce any:

- [x] temporary files, auto-generated files or secret keys
- [x] build works
- [x] eslint issues
- [x] typescript issues
- [x] codegen errors
- [x] `console.log` meant for debugging
- [x] typos
- [x] unwanted comments
- [x] conflict markers

subinasr committed:
Add and setup rest request

AdityaKhatri committed:
Add export functionality

AdityaKhatri committed:
Set map to include data from lowest to highest instead of 0 to highest

tnagorra commented on package.json:
```diff
@@ -40,12 +40,15 @@
+        "@types/papaparse": "^5.3.5",
         "core-js": "3.22.3",
-        "csv-stringify": "5.6.5",
+        "csv-parse": "^5.3.1",
+        "csv-stringify": "^6.2.0",
         "file-saver": "^2.0.5",
         "graphql": "^16.2.0",
         "html-ellipsis": "^2.0.4",
         "mapbox-gl": "^2.8.2",
+        "papaparse": "^5.3.2",
```
Do we use csv-parse, csv-stringify and papaparse?

tnagorra replied:
Done!

tnagorra commented on app/views/Dashboard/Overview/MapView/index.tsx:
```diff
@@ -320,6 +320,7 @@ function MapView(props: Props) {
 
     const formatOnMap = mapIndicatorState[0]?.format ?? 'percent';
 
+    const lowestDataOnMap = formatOnMap === 'percent' ? 0 : mapIndicatorState[mapIndicatorState.length - 1]?.value;
```
What happens if lowestDataOnMap is undefined?

tnagorra replied:
Fixed

tnagorra commented on app/views/Dashboard/index.tsx:
```diff
@@ -310,6 +404,48 @@ function Dashboard() {
     ]);
 
+    useEffect(() => {
+        if (!pendingExport) {
+            if (exportData?.length > 0) {
+                if (exportData.length === exportTotal) {
+                    /*
+                    const dataString = stringify(exportData, {
+                        columns: headers,
+                    });
+                    */
+                    const blob = new Blob(
+                        [exportFullString],
+                        { type: 'text/csv' },
+                    );
+                    saveAs(blob, 'Data Export');
+                } else {
+                    // eslint-disable-next-line no-console
+                    console.error('CSV num rows mismatch', `expected: ${exportTotal}`, `got: ${exportData.length}`);
+                }
+            }
+        }
```
Too must nested ifs.
Let's do this on onCompleted instead of an useEffect.

tnagorra replied:
done!

subinasr replied:
much*

tnagorra commented on app/views/Dashboard/index.tsx:
```diff
@@ -310,6 +404,48 @@ function Dashboard() {
+            return 0;
+        }
+
+        return Math.round(100 * (exportData?.length / exportTotal) ?? 0) / 100;
```
```suggestion
        return Math.round(100 * ((exportData?.length ?? 0) / exportTotal)) / 100;
```

tnagorra replied:
Done!

tnagorra commented on app/Base/configs/restRequest.ts:
```diff
@@ -0,0 +1,16 @@
+    apiHttps,
+} from '#base/configs/env';
+
+export const reactAppApiHttps = location.protocol === 'https:' // eslint-disable-line no-restricted-globals
+    ? 'https'
+    : apiHttps;
+
+export const wsEndpoint = !apiEndpoint
+    ? 'http://localhost:7020/api/v1'
+    : `${reactAppApiHttps}://${apiEndpoint}/api/v1`;
+
+export const adminEndpoint = !apiEndpoint
+    ? 'http://localhost:7020/admin/'
+    : `${reactAppApiHttps}://${apiEndpoint}/admin/`;
```
Why do we even need these?

tnagorra commented on app/Base/utils/restRequest.ts:
```diff
@@ -0,0 +1,57 @@
+}
+
+export const prepareUrl = (url: string) => {
+    if (url.startsWith(serverPrefix)) {
+        // NOTE: -1 to leave out the starting slash
+        const cleanedUrl = url.slice(serverPrefix.length - 1);
+        const a = getVersionedUrl(wsEndpoint, cleanedUrl);
+        return a;
+    }
```
This was a custom code for DEEP. We do not need this here.

tnagorra committed:
Use onCompleted instead of useEffect for useRecursiveCSVExport

- Add headers on csv export
- Remove usage of intermediate string representation for csv
- Use Papa.unparse to create csv
- Use skipEmptyLines instead of using pop on last data point
- Fix issue with extra request call when page size is a factor to total
  pages
- Decrease pagesize for debug purposes
